DataDog/datadog-agent
# [CWS] Tweak CWS network to prevent a race with Cilium

### What does this PR do?



This PR introduces the ability to configure the priority and handle of the eBPF Traffic Control classifiers used by Cloud Workload Security to implement its network detections. Additionally, the default priority level is now set to 10 (instead of 1) to prevent any conflict with Cilium.

Please note that with this change, the network detections from CWS will no longer work inside containers managed by Cilium, until Cilium changes the way they use TC classifiers as well.

### Motivation



This PR is needed as Cilium is currently configured to own Filter 1 and Handle 0x1 on the host side of every lxc veth pair. As loading eBPF TC filters is racy by design (sometimes Cilium is faster than the Datadog Agent, sometimes it is the opposite), it is possible that the Datadog Agent is faster to attach to host side of the veth pair. In such cases, the kernel grants us the 0x1 Handle on prio 1, and Cilium will eventually replace the Datadog Agent TC filter. Then, as CWS checks at runtime if its filters are still active, we will detect that our filter was deleted and we call `Probe.Stop()` on our TC filters to cleanup behind us ... this will eventually delete Cilium's filter.

When this race happens, the new pod will lose all network connectivity. The only way to recover from this race is to restart the pod or the Cilium agent.

### How to test the changes

- Start CWS and schedule a pod in a Kubernetes cluster managed by Cilium.
- Make sure that all Traffic Control classifiers loaded by the Datadog Agent are set with a default priority of 10. To do so, run `tc filter show dev [any_lxc_interface] egress` for example.
